### PR TITLE
[wxDEBUG_COLLECT_TRACE_MASKS=1] dump the collected log mask IDs to the debug channel once the application terminates

### PR DESCRIPTION
**Pull request by GerHobbelt**
tweak: dump the collected log mask IDs to the debug channel once the application terminates, when running in debug mode, when the `wxDEBUG_COLLECT_TRACE_MASKS` flag has been turned ON(1) in the wxWidgets library build settings. 

This is useful for detecting which log masks are employed in a larger application, where static analysis might be more costly (but would provide a more complete picture).

---

IMHO, this is a useful end to the `wxDEBUG_COLLECT_TRACE_MASKS` work that's already present in the wxWidgets core.

### CODE REVIEW

**Comment by vadz** [comment]
This seems to be incomplete, have you forgotten to include the changes to the other files by chance?

In any case, I don't think we want a separate build option for this, I'd rather check for some `WXTRACE_DEBUG` environment variable being defined and dumping the masks when it is.

**Comment by GerHobbelt** [comment]
😓 My mistake. Added the missing code changes for completeness sake.

Somehow the thought had lodged in my mind that this was a continuation of work already present in mainline, **but it is not**: this feature was added by me while I was busy hunting down the cause(s) why my executables turned out *huge* when compiling wxWidgets (and some other libraries) in a monolithic build and how to produce lean binaries when adjusting the various `wxUSE_XYZ` settings, e.g. `wxUSE_LOG`  (all done in static build, debug mode, custom rig). 

> I get some *very* strange entries linked in (says the MSVC linker MAP file, matching the large EXE size) that are used by absolutely *no-body*, but they only happen when some `wxUSE_XYZ` features are on. Still drilling down on that set and hunting the cause. Probably a static instantiation plus dependency cycle somewhere, but I don't know yet -- this is just a hunch while the recompile/debug/diag cycle continues. 🕵️ 

Then, with wxUSE_LOG=1, but some other wxUSE_XYZ settings turned off (0), log/diag channel output wasn't behaving as I'ld expected/hoped, and it turned out, in my case, that the log *mask* filter `wxLog::IsAllowedTraceMask()` was throwing a spanner in the works: the above feature was added then to help diagnose what got *to* the logger but didn't get *through* at run-time.

---

re your `TRACE_DEBUG` comment: I assume you mean one of the `wxLogTrace()` mask `#define`s (or a new one added to that set)? If yes, then I don't see much of a problem. 

> Only one nitpick in that case: `wxLogTrace(mask, ...)` mask constants are a bit of a mixed bunch (direct literal, like "dll"; `wxTRACE_XYZ` define's like `wxTRACE_FSWATCHER`, others don't have `wx`, e.g. `TRACE_KEYS` or `FILECONF_TRACE_MASK`) so I liked the more consistent `wxUSE_XYZ` constants more, also while those are clearly meant to turn features on and off, which is the goal here with `wxDEBUG_COLLECT_TRACE_MASKS`. 😉 
>

🤔 The masks seem more like a run-time switchable thing to me, contrasting the `wxUSE_XYZ`, etc., which are compile-time switchable. If you still want this, do you feel this should rather be a run-time switchable feature? (Personally, I'ld choose compile-time as this is only intended for debugging, but it would make for a different dev cycle. Me, I'm more of a `TRACE=*` wildcard-means-accept-all guy with logging and doing the filtering afterwards and/or externally.)

---

Anyway, bit of a mess, this one, and if you don't want it after all, no harm done. I just added the rest of the code and the above notes so you could at least understand what's intended and has been going on. Sorry for the bother.

**Comment by vadz** [comment]
Sorry, I don't know what's going on with the binaries size, it might make more sense to discuss this on wx mailing lists.

As for the trace masks, they're quite orthogonal to `wxUSE_XXX`, and I agree that it could be useful to have the functionality you want, i.e. to be able to show all the trace masks actually used. However we don't want to recompile the library just to see this -- and neither do we want to have them shown every time the application runs. So we need a run-time switch here if we want to have this at all.

To summarize, this would be acceptable if there were some way to enable it externally _without_ recompiling the library (because nobody will do this anyhow), but not in the current form I'm afraid.

**Comment by GerHobbelt** [comment]
Understood. I'll think of something...

**Comment by vadz** [comment]
This PR has been in "work needed" state for more than a year and will be closed soon. Please update it, e.g. by commenting here, if anybody intends to resume working on it.
